syntax = "proto3";

package SpravSerialize;

option cc_enable_arenas = true;

message Bus {
  uint32 id = 1;
  string name = 2;

  uint32 stops_count = 5;
  uint32 unique_stops_count = 6;
  uint32 length = 7;
  float curvature = 8;
}

message Stop {
  uint32 id = 1;
  string name = 4;

  repeated uint32 bus = 5;
}

message Graph {
  message Edge {
    enum Type {
      NOOP = 0;
      BUS = 1;
      WAIT = 2;
    }

    uint32 from = 1;
    uint32 to = 2;
    float weight = 3;
    Type type = 4;
    uint32 id = 5;
    uint32 span_count = 6;
    repeated uint32 stops = 7;
  }

  repeated Edge edge = 1;
  uint32 vertex_count = 2;
}

message Router {
  message Data {
    bool is_defined = 1;
    float weight = 2;
    bool has_prev_edge = 3;
    uint32 prev_edge = 4;
  }

  message Row {
    repeated Data data = 1;
  }

  repeated Row row = 1;
}

message TransportCatalog {
  repeated Bus bus = 1;
  repeated Stop stop = 2;
  Graph graph = 3;
  Router router = 4;
}